data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/s3/S3ObjectWorker.java

```diff
@@ -64,6 +67,81 @@ public S3ObjectWorker(final S3ObjectRequest s3ObjectRequest) {
         this.s3ObjectPluginMetrics = s3ObjectRequest.getS3ObjectPluginMetrics();
     }
 
+    private void processObjectMetadata(final AcknowledgementSet acknowledgementSet,
+                               final S3ObjectReference s3ObjectReference,
+                               final BufferAccumulator<Record<Event>> bufferAccumulator,
+                               final SourceCoordinator<S3SourceProgressState> sourceCoordinator,
+                               final String partitionKey) throws IOException {
+        final S3InputFile inputFile = new S3InputFile(s3Client, s3ObjectReference, bucketOwnerProvider, s3ObjectPluginMetrics);
+        final String BUCKET = "bucket";
+        final String KEY = "key";
+        final String TIME = "time";
+        final String LENGTH = "length";
+        Map<String, Object> data = new HashMap<>();
+        data.put(BUCKET, s3ObjectReference.getBucketName());
+        data.put(KEY, s3ObjectReference.getKey());
+        data.put(TIME, inputFile.getLastModified());
+        data.put(LENGTH, inputFile.getLength());
+        Event event = JacksonEvent.builder()
+                .withEventType("event")
+                .withData(data)
+                .build();
+        final long s3ObjectSize = event.toJsonString().length();
+        if (acknowledgementSet != null) {
+            acknowledgementSet.add(event);
+        }
+        AtomicLong lastCheckpointTime = new AtomicLong(System.currentTimeMillis());
+        final AtomicInteger saveStateCounter = new AtomicInteger();
+        final Instant lastModifiedTime = inputFile.getLastModified();
+        final Instant now = Instant.now();
+        final Instant originationTime = (lastModifiedTime == null || lastModifiedTime.isAfter(now)) ? now : lastModifiedTime;
+        event.getMetadata().setExternalOriginationTime(originationTime);
+        event.getEventHandle().setExternalOriginationTime(originationTime);
+        try {
+            bufferAccumulator.add(new Record<>(event));
+        } catch (final Exception e) {
+            LOG.error("Failed writing S3 objects to buffer.", e);
+        }
+        if (acknowledgementSet != null && sourceCoordinator != null && partitionKey != null &&
```

Comment:
This seems a very leaky abstraction. Why does this code need to care about all these details? I see we have this in the current code and it is producing duplicate code and unclear responsibilities. We can refactor this with a simpler `Consumer` of some sort to finish the record. Let the S3 scan take care of this.